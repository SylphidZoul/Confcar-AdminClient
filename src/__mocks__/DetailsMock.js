export const detailedDayMock = {
  day_date: '10/8/2020',
  day_end: '18:00',
  day_start: '08:10',
  extraPause_end: '11:20',
  extraPause_start: '10:20',
  fullname: 'Juan',
  lunch_end: '13:50',
  lunch_start: '13:00',
  rawHours: 'PT28800S',
  week: 41,
  workedHours: '08:00'
}
