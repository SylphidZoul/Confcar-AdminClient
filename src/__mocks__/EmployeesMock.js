export const EmployeesData = [
  {
    dni: 38098706,
    employee_id: 2,
    fullname: 'Juan',
    hourly_pay: 110,
    mobile: 1164329741,
    password: 'asd1234'
  },
  {
    dni: 17753281,
    employee_id: 16,
    fullname: 'Carlos',
    hourly_pay: 104,
    mobile: 1164329741,
    password: '852123'
  },
  {
    dni: 37098750,
    employee_id: 54,
    fullname: 'Mariano',
    hourly_pay: 78,
    mobile: 1164329700,
    password: '963741'

  }
]

export const fetchedInitialData = {
  data: {
    body: EmployeesData
  }
}
