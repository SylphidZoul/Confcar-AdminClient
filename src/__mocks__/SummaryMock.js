export const employeeDataMock = [
  {
    detailedDays: [],
    fullname: 'Raúl',
    rawHours: 'PT23400S',
    weekHours: '06:30',
    weekNumber: 41,
    weekPay: 449
  },
  {
    detailedDays: [],
    fullname: 'Agustina',
    rawHours: 'PT234003',
    weekHours: '06:00',
    weekNumber: 41,
    weekPay: 50
  }
]

export const totalDataMock = {
  hours: '12:30',
  pay: 499,
  weekNumber: 41
}

export const fetchedInitialData = {
  data: {
    body: {
      detailsPerEmployee: employeeDataMock,
      ...totalDataMock
    }
  }
}

export const fetchedSelectedDate = {
  data: {
    body: {
      detailsPerEmployee: employeeDataMock.map(employee => {
        return { ...employee, weekNumber: 31 }
      }),
      hours: '12:30',
      pay: 499,
      weekNumber: 31
    }
  }
}

export const fetchedYearData = {
  data: {
    body: {
      detailsPerEmployee: [
        {
          fullname: 'Raúl',
          rawHours: 'PT23400S',
          hours: '06:30',
          pay: 449
        },
        {
          fullname: 'Agustina',
          rawHours: 'PT234003',
          hours: '06:00',
          pay: 50
        }
      ],
      hours: '12:30',
      pay: 499,
      periodNumber: 2020
    }
  }
}
