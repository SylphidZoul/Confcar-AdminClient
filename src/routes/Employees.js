import { useEffect, useReducer } from 'react'
import { employeesReducer, initialState, actions } from '../reducers/employeesReducer'
import Http from '../libs/http'
import EmployeesScreen from '../components/employeesScreen'

const Employees = () => {
  const [state, dispatch] = useReducer(employeesReducer, initialState)

  const fetchEmployees = async () => {
    dispatch({ type: actions.FETCH_DATA })
    const query = state.showInactives ? 'employees/active=0' : 'employees'
    try {
      const { body } = await Http.instance.get(query)
      dispatch({ type: actions.SET_EMPLOYEE_LIST, payload: body })
    } catch (error) {
      console.log('error fetch')
      dispatch({ type: actions.FAILED_CONNECTION })
    }
  }

  const toggleEdit = (index = -1) => {
    if (index === -1) {
      dispatch({ type: actions.DISABLE_EDIT_MODE })
      return
    }
    dispatch({ type: actions.ENABLE_EDIT_MODE, payload: index })
  }

  const handleSubmit = async (e, id) => {
    e.preventDefault()
    const update = {
      id: id,
      fullname: e.target[0].value,
      dni: e.target[1].value,
      password: e.target[2].value,
      mobile: e.target[3].value,
      hourly_pay: e.target[4].value
    }
    toggleEdit()
    dispatch({ type: actions.FETCH_DATA })
    try {
      const updated = await Http.instance.put('employees', update)
      const index = state.employeesList.findIndex(elem => (
        elem.employee_id === updated.body.employee_id
      ))
      const list = [...state.employeesList]
      list[index] = updated.body
      dispatch({ type: actions.SET_EMPLOYEE_LIST, payload: list })
    } catch (error) {
      console.log(error)
      dispatch({ type: actions.FAILED_CONNECTION })
    }
  }

  const handleDelete = async (id) => {
    dispatch({ type: actions.FETCH_DATA })
    try {
      await Http.instance.delete(`employees/${id}`)
      const filtered = state.employeesList.filter(elem => (
        elem.employee_id !== id
      ))
      dispatch({ type: actions.SET_EMPLOYEE_LIST, payload: filtered })
    } catch (error) {
      console.log(error)
      dispatch({ type: actions.FAILED_CONNECTION })
    }
  }

  const toggleInactives = () => {
    dispatch({ type: actions.TOGGLE_SHOW_INACTIVES })
  }

  useEffect(() => {
    fetchEmployees()
  }, [state.showInactives])

  return (
    <EmployeesScreen
      state={state}
      toggleEdit={toggleEdit}
      onSubmit={handleSubmit}
      onDelete={handleDelete}
      toggleInactives={toggleInactives}
    />
  )
}

export default Employees
