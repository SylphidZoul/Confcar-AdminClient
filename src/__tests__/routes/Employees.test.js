import { mount } from 'enzyme'
import { HashRouter } from 'react-router-dom'
import toJson from 'enzyme-to-json'
import { act } from 'react-dom/test-utils'
import Employees from '../../routes/Employees'
import axios from 'axios'
import { fetchedInitialData } from '../../__mocks__/EmployeesMock'

jest.mock('axios')

describe('Employees Tests', () => {
  let wrapper

  beforeEach(async () => {
    axios.get.mockResolvedValueOnce(fetchedInitialData)
    wrapper = mount(
      <HashRouter>
        <Employees />
      </HashRouter>
    )
    expect(wrapper.exists).toBeTruthy()
    expect(wrapper.find('h1').text()).toBe('Cargando...')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(axios.get).toHaveBeenCalledTimes(1)
  })

  test('Validate Snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  test('Initial Rendering', () => {
    expect(wrapper.find('h1').text()).toBe('Empleados Activos')
    // testing table structure
    expect(wrapper.find('ul').children().length).toBe(4)
    expect(wrapper.find('li').children().length).toBe(27)
    expect(wrapper.find('p').length).toBe(18)
    expect(wrapper.find('p').at(8).text()).toBe('Carlos')
    expect(wrapper.find('ul button').length).toBe(6)
  })

  test('Inactive Employees not found', async () => {
    axios.get.mockResolvedValueOnce({ data: { body: [] } })
    wrapper.find('button').at(6).simulate('click')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(wrapper.find('h1').text()).toBe('Empleados inactivos')
    expect(wrapper.find('h2').text()).toBe('No se encontraron empleados.')
  })

  test('Enable edit mode', async () => {
    wrapper.find('button').first().simulate('click')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(wrapper.find('form').length).toBe(1)
    expect(wrapper.find('form input').length).toBe(5)
    expect(wrapper.find('p').length).toBe(13)
  })

  test('Disable edit mode', async () => {
    wrapper.find('button').first().simulate('click')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(wrapper.find('form').length).toBe(1)
    wrapper.find('button').at(1).simulate('click')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(wrapper.find('ul').children().length).toBe(4)
    expect(wrapper.find('li').children().length).toBe(27)
    expect(wrapper.find('p').length).toBe(18)
  })

  test('Send an Employee to inactive list', async () => {
    axios.delete.mockResolvedValueOnce({ data: { body: {} } })
    wrapper.find('li button').at(1).simulate('click')
    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0))
    })
    wrapper.update()
    expect(wrapper.find('ul').children().length).toBe(3)
  })
})
