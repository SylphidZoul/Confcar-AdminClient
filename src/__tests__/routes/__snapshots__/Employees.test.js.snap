// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employees Tests Validate Snapshot 1`] = `
<HashRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Employees>
      <EmployeesScreen
        onDelete={[Function]}
        onSubmit={[Function]}
        state={
          Object {
            "editMode": false,
            "employeesList": Array [
              Object {
                "dni": 38098706,
                "employee_id": 2,
                "fullname": "Juan",
                "hourly_pay": 110,
                "mobile": 1164329741,
                "password": "asd1234",
              },
              Object {
                "dni": 17753281,
                "employee_id": 16,
                "fullname": "Carlos",
                "hourly_pay": 104,
                "mobile": 1164329741,
                "password": "852123",
              },
              Object {
                "dni": 37098750,
                "employee_id": 54,
                "fullname": "Mariano",
                "hourly_pay": 78,
                "mobile": 1164329700,
                "password": "963741",
              },
            ],
            "isFetching": false,
            "selectedEmployee": -1,
            "showInactives": false,
          }
        }
        toggleEdit={[Function]}
        toggleInactives={[Function]}
      >
        <styled.h1>
          <h1
            className="sc-fujyUd bhxXcD"
          >
            Empleados Activos
          </h1>
        </styled.h1>
        <EditableTable
          editMode={false}
          fieldList={
            Array [
              "Id",
              "Nombre",
              "DNI",
              "Contraseña",
              "Teléfono",
              "Pesos/Hora",
            ]
          }
          handleDelete={[Function]}
          idKey="employee_id"
          itemList={
            Array [
              Object {
                "dni": 38098706,
                "employee_id": 2,
                "fullname": "Juan",
                "hourly_pay": 110,
                "mobile": 1164329741,
                "password": "asd1234",
              },
              Object {
                "dni": 17753281,
                "employee_id": 16,
                "fullname": "Carlos",
                "hourly_pay": 104,
                "mobile": 1164329741,
                "password": "852123",
              },
              Object {
                "dni": 37098750,
                "employee_id": 54,
                "fullname": "Mariano",
                "hourly_pay": 78,
                "mobile": 1164329700,
                "password": "963741",
              },
            ]
          }
          renderEditableItem={[Function]}
          selectedItem={-1}
          toggleEdit={[Function]}
        >
          <styled.ul>
            <ul
              className="sc-bdnylx eTsEhl"
            >
              <styled.li
                columns={6}
              >
                <li
                  className="sc-gtssRu gJGQeg"
                >
                  <styled.h3
                    key="0"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      Id
                    </h3>
                  </styled.h3>
                  <styled.h3
                    key="1"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      Nombre
                    </h3>
                  </styled.h3>
                  <styled.h3
                    key="2"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      DNI
                    </h3>
                  </styled.h3>
                  <styled.h3
                    key="3"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      Contraseña
                    </h3>
                  </styled.h3>
                  <styled.h3
                    key="4"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      Teléfono
                    </h3>
                  </styled.h3>
                  <styled.h3
                    key="5"
                  >
                    <h3
                      className="sc-hKFyIo etqfMD"
                    >
                      Pesos/Hora
                    </h3>
                  </styled.h3>
                </li>
              </styled.li>
              <TableItem
                handleDelete={[Function]}
                id={2}
                index={0}
                item={
                  Object {
                    "dni": 38098706,
                    "employee_id": 2,
                    "fullname": "Juan",
                    "hourly_pay": 110,
                    "mobile": 1164329741,
                    "password": "asd1234",
                  }
                }
                key="0"
                toggleEdit={[Function]}
              >
                <Styled(styled.li)
                  columns={6}
                  index={0}
                >
                  <li
                    className="sc-gtssRu sc-dlnjPT gJGQeg hcaTng"
                  >
                    <styled.p
                      key="0"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        38098706
                      </p>
                    </styled.p>
                    <styled.p
                      key="1"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        2
                      </p>
                    </styled.p>
                    <styled.p
                      key="2"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        Juan
                      </p>
                    </styled.p>
                    <styled.p
                      key="3"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        110
                      </p>
                    </styled.p>
                    <styled.p
                      key="4"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        1164329741
                      </p>
                    </styled.p>
                    <styled.p
                      key="5"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        asd1234
                      </p>
                    </styled.p>
                    <styled.div>
                      <div
                        className="sc-jSFkmK iTPIHR"
                      >
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Pencil>
                              <svg
                                fill="#329A97"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                                />
                              </svg>
                            </Pencil>
                          </button>
                        </styled.button>
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Thrash>
                              <svg
                                fill="#9B0821"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                                />
                              </svg>
                            </Thrash>
                          </button>
                        </styled.button>
                      </div>
                    </styled.div>
                  </li>
                </Styled(styled.li)>
              </TableItem>
              <TableItem
                handleDelete={[Function]}
                id={16}
                index={1}
                item={
                  Object {
                    "dni": 17753281,
                    "employee_id": 16,
                    "fullname": "Carlos",
                    "hourly_pay": 104,
                    "mobile": 1164329741,
                    "password": "852123",
                  }
                }
                key="1"
                toggleEdit={[Function]}
              >
                <Styled(styled.li)
                  columns={6}
                  index={1}
                >
                  <li
                    className="sc-gtssRu sc-dlnjPT gJGQeg iUxqhc"
                  >
                    <styled.p
                      key="0"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        17753281
                      </p>
                    </styled.p>
                    <styled.p
                      key="1"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        16
                      </p>
                    </styled.p>
                    <styled.p
                      key="2"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        Carlos
                      </p>
                    </styled.p>
                    <styled.p
                      key="3"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        104
                      </p>
                    </styled.p>
                    <styled.p
                      key="4"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        1164329741
                      </p>
                    </styled.p>
                    <styled.p
                      key="5"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        852123
                      </p>
                    </styled.p>
                    <styled.div>
                      <div
                        className="sc-jSFkmK iTPIHR"
                      >
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Pencil>
                              <svg
                                fill="#329A97"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                                />
                              </svg>
                            </Pencil>
                          </button>
                        </styled.button>
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Thrash>
                              <svg
                                fill="#9B0821"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                                />
                              </svg>
                            </Thrash>
                          </button>
                        </styled.button>
                      </div>
                    </styled.div>
                  </li>
                </Styled(styled.li)>
              </TableItem>
              <TableItem
                handleDelete={[Function]}
                id={54}
                index={2}
                item={
                  Object {
                    "dni": 37098750,
                    "employee_id": 54,
                    "fullname": "Mariano",
                    "hourly_pay": 78,
                    "mobile": 1164329700,
                    "password": "963741",
                  }
                }
                key="2"
                toggleEdit={[Function]}
              >
                <Styled(styled.li)
                  columns={6}
                  index={2}
                >
                  <li
                    className="sc-gtssRu sc-dlnjPT gJGQeg hcaTng"
                  >
                    <styled.p
                      key="0"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        37098750
                      </p>
                    </styled.p>
                    <styled.p
                      key="1"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        54
                      </p>
                    </styled.p>
                    <styled.p
                      key="2"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        Mariano
                      </p>
                    </styled.p>
                    <styled.p
                      key="3"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        78
                      </p>
                    </styled.p>
                    <styled.p
                      key="4"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        1164329700
                      </p>
                    </styled.p>
                    <styled.p
                      key="5"
                    >
                      <p
                        className="sc-eCApGN dDzstV"
                      >
                        963741
                      </p>
                    </styled.p>
                    <styled.div>
                      <div
                        className="sc-jSFkmK iTPIHR"
                      >
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Pencil>
                              <svg
                                fill="#329A97"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 000-1.41l-2.34-2.34a.996.996 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                                />
                              </svg>
                            </Pencil>
                          </button>
                        </styled.button>
                        <styled.button
                          onClick={[Function]}
                        >
                          <button
                            className="sc-gKAblj eQkBzg"
                            onClick={[Function]}
                          >
                            <Thrash>
                              <svg
                                fill="#9B0821"
                                height={16}
                                viewBox="0 0 24 24"
                                width={16}
                              >
                                <path
                                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zm2.46-7.12l1.41-1.41L12 12.59l2.12-2.12 1.41 1.41L13.41 14l2.12 2.12-1.41 1.41L12 15.41l-2.12 2.12-1.41-1.41L10.59 14l-2.13-2.12zM15.5 4l-1-1h-5l-1 1H5v2h14V4z"
                                />
                              </svg>
                            </Thrash>
                          </button>
                        </styled.button>
                      </div>
                    </styled.div>
                  </li>
                </Styled(styled.li)>
              </TableItem>
            </ul>
          </styled.ul>
        </EditableTable>
        <Styled(Link)
          to="/employees/signup"
        >
          <Link
            className="sc-jrsJCI hlYSgN"
            to="/employees/signup"
          >
            <LinkAnchor
              className="sc-jrsJCI hlYSgN"
              href="#/employees/signup"
              navigate={[Function]}
            >
              <a
                className="sc-jrsJCI hlYSgN"
                href="#/employees/signup"
                onClick={[Function]}
              >
                Agregar nuevo empleado
              </a>
            </LinkAnchor>
          </Link>
        </Styled(Link)>
        <styled.button
          onClick={[Function]}
        >
          <button
            className="sc-pNWxx heFUkN"
            onClick={[Function]}
          >
            Ver empleados 
            inactivos
          </button>
        </styled.button>
      </EmployeesScreen>
    </Employees>
  </Router>
</HashRouter>
`;
